stack/eslint-plugin-packlets/src/DependencyAnalyzer.ts

```diff
@@ -0,0 +1,142 @@
+interface RefFile {
+  referencedFileName: string;
+  kind: RefFileKind;
+  index: number;
+  file: string;
+}
+
+/**
+ * Represents a packlet that imports another packlet.
+ */
+export interface IPackletImport {
+  /**
+   * The name of the packlet being imported.
+   */
+  packletName: string;
+
+  /**
+   * The absolute path of the file that imports the packlet.
+   */
+  fromFilePath: string;
+}
+
+/**
+ * Used to build a linked list of imports that represent a circular dependency.
+ */
+interface IImportListNode extends IPackletImport {
+  /**
+   * The previous link in the linked list.
+   */
+  previousNode: IImportListNode | undefined;
+}
+
+export class DependencyAnalyzer {
+  private static walkImports(
+    packletName: string,
+    startingPackletName: string,
+    refFileMap: Map<string, RefFile[]>,
+    program: ts.Program,
+    packletsFolderPath: string,
+    visitedPacklets: Set<string>,
+    previousNode: IImportListNode | undefined
+  ): IImportListNode | undefined {
+    const packletEntryPoint: string = path.join(packletsFolderPath, packletName, 'index');
+
+    const tsSourceFile: ts.SourceFile | undefined =
+      program.getSourceFile(packletEntryPoint + '.ts') || program.getSourceFile(packletEntryPoint + '.tsx');
+    if (!tsSourceFile) {
+      return undefined;
+    }
+
+    const refFiles: RefFile[] | undefined = refFileMap.get((tsSourceFile as any).path as any);
+    if (!refFiles) {
+      return undefined;
+    }
+
+    for (const refFile of refFiles) {
+      if (refFile.kind === RefFileKind.Import) {
+        const referencingFilePath: string = refFile.file;
+
+        if (Path.isUnder(referencingFilePath, packletsFolderPath)) {
+          const referencingRelativePath: string = path.relative(packletsFolderPath, referencingFilePath);
+          const referencingPathParts: string[] = referencingRelativePath.split(/[\/\\]+/);
+          const referencingPackletName: string = referencingPathParts[0];
+
+          if (!visitedPacklets.has(packletName)) {
+            visitedPacklets.add(packletName);
+
+            const importListNode: IImportListNode = {
+              previousNode: previousNode,
+              fromFilePath: referencingFilePath,
+              packletName: packletName
+            };
+
+            if (referencingPackletName === startingPackletName) {
+              return importListNode;
+            }
+
+            const result: IImportListNode | undefined = DependencyAnalyzer.walkImports(
+              referencingPackletName,
+              startingPackletName,
+              refFileMap,
+              program,
+              packletsFolderPath,
+              visitedPacklets,
+              importListNode
+            );
+            if (result) {
+              return result;
+            }
+          }
+        }
+      }
+    }
+
+    return undefined;
+  }
+
+  public static detectCircularImport(
```

Comment:
I assume we want to return only the first cycle if there are multiple cycles?

Reply:
I should add some docs to clarify this.  The way it works:

- Cycles are only detected by looking at `index.ts` files. This works because we require consumers to import those files.
- Each `import` statement is analyzed separately, and the analysis stops when the first cycle is found.
- We then sort the list of affected packlets alphabetically, and only report the error for the `index.ts` file of the first packlet in the list.  This avoids redundantly reporting it for each packlet along the cycle.